File: universal-adapter/src/main/java/com/jacekmarchwicki/universaladapter/BaseViewHolderManager.java
package com.jacekmarchwicki.universaladapter;

import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import javax.annotation.Nonnull;

public class BaseViewHolderManager<T extends BaseAdapterItem> implements ViewHolderManager {

    public interface ViewHolderFactory<TT extends BaseAdapterItem> {
        @NonNull BaseViewHolder<TT> createViewHolder(@NonNull View view);
    }

    @NonNull
    private final Class<? extends T> clazz;
    @LayoutRes
    private final int mLayoutRes;
    @NonNull
    private final ViewHolderFactory<T> mViewHolderFactory;

    public BaseViewHolderManager(@LayoutRes int layoutRes, @NonNull ViewHolderFactory<T> viewHolderFactory, @NonNull Class<? extends T> clazz) {
        this.clazz = clazz;
        mLayoutRes = layoutRes;
        mViewHolderFactory = viewHolderFactory;
    }

    @Override
    public boolean matches(@Nonnull BaseAdapterItem baseAdapterItem) {
        return clazz.isInstance(baseAdapterItem);
    }

    @Nonnull
    @Override
    public BaseViewHolder createViewHolder(@Nonnull ViewGroup viewGroup, @Nonnull LayoutInflater layoutInflater) {
        final View itemView = layoutInflater.inflate(mLayoutRes, viewGroup, false);
        return mViewHolderFactory.createViewHolder(itemView);
    }
}

File: universal-adapter-kotlin/src/main/java/com/appunite/universal_adapter_kotlin/BaseViewHolderManager.kt
package com.appunite.universal_adapter_kotlin

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import com.jacekmarchwicki.universaladapter.BaseAdapterItem
import kotlin.reflect.KClass

abstract class BaseViewHolderManager<T : BaseAdapterItem>(
        private val viewHolderCreator: (view: View) -> BaseViewHolder<T>,
        private val clazz: KClass<out T>
) : ViewHolderManager {

    abstract fun createView(parent: ViewGroup, inflater: LayoutInflater): View

    override fun createViewHolder(
            parent: ViewGroup,
            inflater: LayoutInflater
    ): BaseViewHolder<BaseAdapterItem> {
        return viewHolderCreator(createView(parent, inflater)) as BaseViewHolder<BaseAdapterItem>
    }

    override fun matches(baseAdapterItem: BaseAdapterItem): Boolean =
            clazz.isInstance(baseAdapterItem)
}

open class LayoutViewHolderManager<T : BaseAdapterItem>(
        @LayoutRes private val layoutRes: Int,
        viewHolderCreator: (view: View) -> BaseViewHolder<T>,
        clazz: KClass<out T>
) : BaseViewHolderManager<T>(viewHolderCreator, clazz) {

    override fun createView(parent: ViewGroup, inflater: LayoutInflater): View {
        return inflater.inflate(layoutRes, parent, false)
    }
}

object ViewHolderManagerFactory {

    inline fun <reified T : BaseAdapterItem> create(
            @LayoutRes layoutRes: Int,
            noinline viewHolderFactory: (View) -> BaseViewHolder<T>
    ): ViewHolderManager {
        return LayoutViewHolderManager(
                layoutRes,
                viewHolderFactory,
                T::class
        )
    }
}